File: targets/widget/Live.swift
import Foundation

/// The widget's own poll of `GET /cronogol/live` (ADR 0080).
///
/// ⚠⚠ **This is the ONE sanctioned network call in this target, and it is
/// rationed.** ADR 0025 step 4's "no network in a timeline provider" was about
/// crest artwork and stands for everything else; 0080 carves out exactly this
/// route, exactly inside a match window — some snapshot row kicked off within
/// the last 150 minutes. Outside the window `FixtureProvider` never calls this,
/// so a widget with nothing in play spends nothing.
///
/// ⚠ The reload cadence while live is 5 minutes (`Cadence.liveReload`): ~30
/// reloads across a match with stoppage, against WidgetKit's ~40–70/day budget.
/// The minute STILL moves every minute — per-minute timeline ENTRIES are free
/// (ADR 0047 "entries are free, reloads are rationed") and tick the anchor in
/// `WidgetLiveState.Match.displayMinute(at:)` on-device. Goal pushes reload on
/// top, through the notification service extension.
enum LiveFetch {
  /// The wire shape — only what the ledger draws. `LiveMatchView` in
  /// `src/lib/cronogol/types.ts` is the full contract; unknown fields are
  /// ignored by `JSONDecoder`, which is the forward-compatibility here.
  struct RouteMatch: Decodable {
    struct Score: Decodable {
      let home: Int?
      let away: Int?
    }
    struct Event: Decodable {
      struct Person: Decodable { let name: String }
      let type: String?
      let minute: Int?
      let player: Person?
    }
    let fixtureId: String
    let status: String
    /// ⚠ Null unless live; includes stoppage (`94` = 90+4). Never render `0′`.
    let minute: Int?
    let score: Score
    /// ⚠ `[]` does not distinguish "nothing happened" from "not fetched yet"
    /// (LIVE-SCORES.md §2) — the scorer line is a bonus, never load-bearing.
    /// Optional: the field predates its own deployment (backend §98).
    let events: [Event]?
  }

  private struct Response: Decodable {
    let matches: [RouteMatch]
  }

  /// ⚠ `?league=laliga` — the API slug, never our `la-liga` (the same trap
  /// `queries/keys.ts` documents). Any OTHER parameter is a 400.
  private static let url = URL(string: "https://crono-gol.com/cronogol/live?league=laliga")!

  /// Nil on any failure — offline, timeout, non-200, undecodable. The caller
  /// keeps whatever `live.json` already holds; a failed poll must never blank
  /// a score that was true five minutes ago.
  static func fetch() async -> [RouteMatch]? {
    var request = URLRequest(url: url)
    // ⚠ Short. A timeline provider has seconds, not the app's patience — a
    // hung request here delays every widget on the home screen.
    request.timeoutInterval = 4

    guard
      let (data, response) = try? await URLSession.shared.data(for: request),
      let http = response as? HTTPURLResponse,
      (200..<300).contains(http.statusCode),
      let decoded = try? JSONDecoder().decode(Response.self, from: data)
    else { return nil }

    return decoded.matches
  }
}

enum LiveMerge {
  /// Fold a fresh poll into the persisted state.
  ///
  /// The rules, in order, per snapshot row that is in the in-play window:
  /// - **Route serves it live** → `live`, minute re-anchored — except the
  ///   half-time reads: a minute frozen at 45 across two polls, or a null
  ///   minute long after kickoff, both mean the break (`halfTime`). The route
  ///   has no break state to give us (LIVE-SCORES.md), so this is a heuristic
  ///   and recorded as one in ADR 0080.
  /// - **Route no longer serves a row we HAVE seen live** → `finished`, last
  ///   score kept. Rows vanish at the whistle; the vanishing IS full time.
  /// - **Route has never served it** → nothing. The ledger draws the
  ///   kicked-off dashes from the snapshot row alone (ADR 0078's tier, on the
  ///   widget), and fabricating a state row for it would turn "no data yet"
  ///   into "finished" on the next merge.
  static func merge(
    previous: WidgetLiveState?,
    route: [LiveFetch.RouteMatch],
    inPlay: [WidgetSnapshot.Entry],
    now: Date
  ) -> WidgetLiveState {
    let served = Dictionary(uniqueKeysWithValues: route.map { ($0.fixtureId, $0) })
    var matches: [WidgetLiveState.Match] = []

    for row in inPlay {
      let old = previous?.match(for: row.fixtureId)

      // ⚠ The `_debug/widgets` sample writer's escape hatch: a fabricated
      // `debug-` fixture is not on the route, and without this the merge
      // (correctly) declares it finished on the very first poll — which is how
      // the FT path got verified, and also why the LIVE preview needs the
      // guard. Real fixture ids are UUIDs; nothing else collides.
      if row.fixtureId.hasPrefix("debug-") {
        if let old { matches.append(old) }
        continue
      }

      if let fresh = served[row.fixtureId], fresh.status == "live" {
        let minute = fresh.minute == 0 ? nil : fresh.minute
        let sinceKickoff = now.timeIntervalSince(row.kickoffUtc)

        // The break, read two ways — see the header.
        let frozenAt45 = minute == 45 && old?.status != "finished" && old?.minute == 45
          && (old.map { now.timeIntervalSince($0.minuteAt) > 3 * 60 } ?? false)
        let nullAfterAnHour = minute == nil && sinceKickoff > 50 * 60

        let lastGoal = (fresh.events ?? []).last { $0.type == "goal" }
        let lastEvent = lastGoal.flatMap { event -> String? in
          guard let name = event.player?.name, let at = event.minute else { return nil }
          return "\(name) \(at)′"
        }

        matches.append(WidgetLiveState.Match(
          fixtureId: row.fixtureId,
          status: (frozenAt45 || nullAfterAnHour) ? "halfTime" : "live",
          minute: minute,
          // ⚠ Keep the OLD anchor when the reported minute has not moved —
          // re-anchoring an unchanged 45 every poll is what would make the
          // frozen-minute read above never fire.
          minuteAt: (old?.minute == minute ? old?.minuteAt : nil) ?? now,
          homeGoals: fresh.score.home,
          awayGoals: fresh.score.away,
          lastEvent: lastEvent ?? old?.lastEvent
        ))
        continue
      }

      if var old, old.isLive {
        // Seen live, now gone: full time. The score is the last one served.
        old.status = "finished"
        old.minute = nil
        matches.append(old)
        continue
      }

      if let old { matches.append(old) } // Already finished — hold it.
    }

    return WidgetLiveState(v: 1, writtenAt: now, source: "widget", matches: matches)
  }
}

File: src/features/widgets/live.ts
/**
 * The widget's live sidecar — `widget/live.json` in the App Group (ADR 0080).
 *
 * ⚠⚠ **Three writers, one shape.** This file is also written by the widget
 * extension itself (its 5-minute poll of `/cronogol/live`, `targets/widget/
 * Live.swift`) and by the notification service extension (a goal / red card /
 * full-time push, `NotificationService.swift`). `targets/_shared/LiveState.swift`
 * is the Swift mirror; a field added to one side is invisible until it is in
 * both. Keep the shape SMALL — it exists so a separate process can draw a
 * score, not to mirror `LiveMatchView`.
 *
 * ⚠ This writer runs only while the app is FOREGROUND with the Today board's
 * poll running — it keeps the widget in step with what the reader is looking
 * at. With the app closed the widget's own poll and the pushes carry it.
 *
 * ⚠ **Never writes an empty `matches` list.** The file expires by arithmetic —
 * the extension only joins rows whose kickoff is inside the 150-minute hold
 * (or, `finished`, the same local day) — so clearing it buys nothing, and an
 * empty overwrite would erase a `finished` entry another writer is holding for
 * the full-time card.
 */
import { reloadWidgets } from '@/features/push/capability';
import { liveMinute } from '@/lib/cronogol/live';
import type { LiveMatchView } from '@/lib/cronogol/types';

import { SNAPSHOT_DIR, writeGroupJson } from './snapshot';

export const LIVE_NAME = 'live.json';

/** Bump when the shape changes. `LiveState.swift` tolerates all versions. */
export const LIVE_VERSION = 1;

export interface WidgetLiveMatch {
  fixtureId: string;
  /**
   * `halfTime` is DERIVED, not served — `/cronogol/live` has no break state.
   * This writer emits it when the minute is null though the match kicked off
   * long ago; the widget's own poll adds a "minute stopped moving at 45"
   * heuristic. `finished` comes from the other two writers only — a row
   * vanishing from the route (widget poll) or a full-time push.
   */
  status: 'live' | 'halfTime' | 'finished';
  /** The minute AS REPORTED, stoppage folded in (`94` = 90+4). Never 0. */
  minute: number | null;
  /**
   * When `minute` was true, so the extension can tick it on-device between
   * writes — the same synthetic-anchor idea as the Live Activity's
   * `clockFromEpoch` (`MatchAttributes.swift`), kept as an instant + minute
   * pair because this file, unlike ActivityKit's payload, is ours to decode.
   */
  minuteAt: string;
  /** ⚠ Null means "no score yet", never nil–nil — render a dash. */
  homeGoals: number | null;
  awayGoals: number | null;
  /** `Iglesias 20′` — the last goal, already rendered. Null until events land. */
  lastEvent: string | null;
}

export interface WidgetLiveState {
  v: number;
  writtenAt: string;
  /** Who wrote last — `app` | `widget` | `push`. Diagnostic only. */
  source: string;
  matches: WidgetLiveMatch[];
}

/**
 * Minutes since kickoff past which a null minute reads as the BREAK rather
 * than as "kicked off, no data yet". Generous: first-half stoppage has run to
 * ~12 minutes in LaLiga.
 */
const HALF_TIME_AFTER_MS = 50 * 60 * 1000;

/** Pure — the same testability split as `buildSnapshot`. */
export function buildWidgetLive(
  matches: readonly LiveMatchView[],
  now: Date,
): WidgetLiveState {
  const rows: WidgetLiveMatch[] = [];

  for (const match of matches) {
    if (match.status !== 'live') continue;

    const minute = liveMinute(match);
    const kickoff = Date.parse(match.kickoffUtc);
    const atBreak =
      minute === null &&
      !Number.isNaN(kickoff) &&
      now.getTime() - kickoff > HALF_TIME_AFTER_MS;

    // The last goal, as a rendered line. ⚠ `events` is often `[]` while the
    // score is not — a fetch that has not landed, not unattributed goals
    // (LIVE-SCORES.md §2) — so the line is a bonus, never load-bearing.
    const lastGoal = [...match.events].reverse().find((event) => event.type === 'goal');

    rows.push({
      fixtureId: match.fixtureId,
      status: atBreak ? 'halfTime' : 'live',
      minute,
      minuteAt: now.toISOString(),
      homeGoals: match.score.home,
      awayGoals: match.score.away,
      lastEvent: lastGoal ? `${lastGoal.player.name} ${lastGoal.minute}′` : null,
    });
  }

  return {
    v: LIVE_VERSION,
    writtenAt: now.toISOString(),
    source: 'app',
    matches: rows,
  };
}

/**
 * The "did anything move" key. ⚠ Excludes `writtenAt` AND `minuteAt` AND the
 * minute itself — the same budget rule as `snapshotKey`, tightened: the minute
 * advances every poll, and a reload per minute is exactly the spend that
 * freezes the widget for the day. The extension ticks the minute on-device
 * from `minuteAt`; the app only needs to reload on a SCORE or STATE change.
 */
export function widgetLiveKey(state: WidgetLiveState): string {
  return JSON.stringify(
    state.matches.map((match) => [
      match.fixtureId,
      match.status,
      match.homeGoals,
      match.awayGoals,
      match.lastEvent,
    ]),
  );
}

let lastKey: string | null = null;

/**
 * Write and reload, if a score or state actually changed.
 *
 * ⚠ No debounce — the caller is `useLive`'s 15-second poll, which is already
 * the slow path; the key guard is what protects the reload budget. Skips
 * entirely when nothing is live (see the file header for why an empty list is
 * never written).
 */
export function applyWidgetLive(matches: readonly LiveMatchView[], now: Date): void {
  const state = buildWidgetLive(matches, now);
  if (state.matches.length === 0) return;

  const key = widgetLiveKey(state);
  if (key === lastKey) return;

  if (!writeGroupJson(SNAPSHOT_DIR, LIVE_NAME, state)) return;
  lastKey = key;
  void reloadWidgets();
}
